api/src/main/java/org/openmrs/validator/ObsValidator.java

```diff
@@ -252,5 +258,129 @@ else if (!isObsGroup) {
 			}
 		}
 	}
+
+	/**
+	 * This method validates Obs' numeric values:
+	 * <ol>
+	 *     <li>Validates Obs in relation to criteria e.g. checks patient's age is within the valid range</li>
+	 *     <li>Validates if Obs' numeric value is within the valid range; i.e. >= low absolute && <= high absolute.</li>
+	 *     <li>Sets field errors if numeric value is outside the valid range</li>
+	 * <ol/>
+	 *
+	 * @param obs Observation to validate
+	 * @param errors Errors to record validation issues
+	 *               
+	 * @since 2.7.0
+	 */
+	private void validateConceptReferenceRange(Obs obs, Errors errors) {
+		ConceptReferenceRange conceptReferenceRange = evaluateReferenceRange(obs.getConcept(), obs.getPerson());
+
+		if (conceptReferenceRange != null) {
+			validateAbsoluteRanges(obs, conceptReferenceRange, errors);
+			setObsReferenceRange(obs, conceptReferenceRange);
+		}
+	}
+
+	/**
+	 * Evaluates the criteria and return the most strict {@link ConceptReferenceRange} for a given concept.
+	 * It considers all valid ranges that match the criteria for the person.
+	 *
+	 * @param concept The concept to evaluate
+	 * @param person The person for whom the range is being evaluated
+	 * @return The strictest {@link ConceptReferenceRange}, or null if no valid range is found
+	 * 
+	 * @since 2.7.0
+	 */
+	public static ConceptReferenceRange evaluateReferenceRange(Concept concept, Person person) {
+		if (concept != null && concept.getDatatype() != null && concept.getDatatype().isNumeric()) {
```

Comment:
To reduce on the level of nesting, can we just return when this condition evaluates to false?